iroh-dns-server/src/store.rs

```diff
@@ -0,0 +1,110 @@
+use std::path::Path;
+
+use anyhow::Result;
+use bytes::Bytes;
+use iroh_metrics::inc;
+use pkarr::{PublicKey, SignedPacket};
+use redb::{backends::InMemoryBackend, Database, ReadableTable, TableDefinition};
+
+use crate::metrics::Metrics;
+
+type PublicKeyBytes = [u8; 32];
+
+const SIGNED_PACKETS_TABLE: TableDefinition<&PublicKeyBytes, &[u8]> =
+    TableDefinition::new("signed-packets-1");
+
+pub struct SignedPacketStore {
+    db: Database,
+}
+
+impl SignedPacketStore {
+    pub fn open_file(path: impl AsRef<Path>) -> Result<Self> {
+        let db = Database::builder().create(path)?;
+        Self::open(db)
+    }
+
+    pub fn open_inmemory() -> Result<Self> {
+        let db = Database::builder().create_with_backend(InMemoryBackend::new())?;
+        Self::open(db)
+    }
+
+    pub fn open(db: Database) -> Result<Self> {
+        let write_tx = db.begin_write()?;
+        {
+            let _table = write_tx.open_table(SIGNED_PACKETS_TABLE)?;
+        }
+        write_tx.commit()?;
+        Ok(Self { db })
+    }
+
+    pub fn upsert(&self, packet: SignedPacket) -> Result<bool> {
+        let key = packet.public_key();
+        let tx = self.db.begin_write()?;
+        let mut inserted = true;
+        {
+            let mut table = tx.open_table(SIGNED_PACKETS_TABLE)?;
+            if let Some(existing) = get_packet(&table, key)? {
+                inserted = false;
+                if existing.more_recent_than(&packet) {
+                    return Ok(false);
+                }
+            }
+            let value = packet.as_bytes();
+            table.insert(&key.to_bytes(), &value[..])?;
+        }
+        tx.commit()?;
```

Comment:
If this will be called many times per second, like during a DOS attempt, committing on every write will be a big bottleneck.

See #3